include/fmt/ranges.h

```diff
@@ -526,20 +530,43 @@ OutputIt write_range_entry(OutputIt out, const Arg& v) {
   return write<Char>(out, v);
 }
 
+#if FMT_HAS_CPP20_CONCEPTS
+template <typename T>
+concept Range = is_range_<T>::value;
+
+template <typename T>
+concept TupleLike = is_tuple_like_<T>::value;
+
+template <typename T>
+concept StdStringLike = is_std_string_like<T>::value;
+#endif
```

Comment:
I don't think introducing concepts is worth it. It just complicates the implementation since we still have to support pre-C++20.

Reply:
I looked at the CI reports and I think the errors on linux and mac are due to my removal of the `detail::write_delimiter` function. The delimiter is now defined in the`fmt::formatting_{tuple, range}` class as `char[8]` and this is passed to `detal::copy()` function to write to the output iterator. However, the original `detail::copy()` does not resolves the overload correctly for char[] to the second `copy(const char*, ...)`, but to the first one `copy(const RangeT&, ...)`. I added some traits check to force the resolution but these traits may not be available in older gcc and clang.

As far as the windows builds, I am not familiar with msvs so I am not sure. Maybe just disable this feature for windows.

 

> I don't think introducing concepts is worth it. It just complicates the implementation since we still have to support pre-C++20.
On the other hand I do think the ability to override existing partial specialization of the fmt::formatter is an important feature for practical use of custom formatting. I tends to think having to maintain a single way of formatting certain types within the fmtlib (ranges.h, chrono.h, etc) is quite harder than allowing custom code to do the thing while fmtlib provides the necessary utility.

Reply:
@vitaut
I've addressed the errors in the CI report.

Reply:
Overriding is fine but it shouldn't rely on C++20 concepts.

Reply:
I am not sure the best way to provide a more specialized template definition for custom formatting of the std range types with SFAINE. I'll look into it when time permit.